File: app/src/main/kotlin/com/foodshare/features/admin/data/SupabaseAdminRepository.kt
package com.foodshare.features.admin.data

import com.foodshare.features.admin.domain.model.AdminAuditLog
import com.foodshare.features.admin.domain.model.AdminDashboardStats
import com.foodshare.features.admin.domain.model.AdminUserProfile
import com.foodshare.features.admin.domain.model.ModerationQueueItem
import com.foodshare.features.admin.domain.model.ModerationResolution
import com.foodshare.features.admin.domain.model.Role
import com.foodshare.features.admin.domain.repository.AdminRepository
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.rpc
import kotlinx.serialization.SerialName
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class SupabaseAdminRepository @Inject constructor(
    private val supabaseClient: SupabaseClient
) : AdminRepository {

    // Dashboard

    override suspend fun fetchDashboardStats(): Result<AdminDashboardStats> = runCatching {
        supabaseClient.postgrest.rpc("get_dashboard_stats")
            .decodeSingle<AdminDashboardStats>()
    }

    // Users

    override suspend fun fetchUsers(
        query: String,
        role: String?,
        limit: Int,
        offset: Int
    ): Result<List<AdminUserProfile>> = runCatching {
        supabaseClient.from("profiles")
            .select {
                filter {
                    if (query.isNotBlank()) {
                        or {
                            ilike("display_name", "%$query%")
                            ilike("email", "%$query%")
                        }
                    }
                }
                limit(limit.toLong())
                range(offset.toLong(), (offset + limit - 1).toLong())
                order("created_at", io.github.jan.supabase.postgrest.query.Order.DESCENDING)
            }
            .decodeList<AdminUserProfile>()
    }

    override suspend fun fetchUser(id: String): Result<AdminUserProfile> = runCatching {
        supabaseClient.from("profiles")
            .select {
                filter { eq("id", id) }
                limit(1)
            }
            .decodeSingle<AdminUserProfile>()
    }

    override suspend fun banUser(userId: String, reason: String): Result<Unit> = runCatching {
        supabaseClient.from("profiles")
            .update(
                buildJsonObject {
                    put("is_banned", true)
                    put("ban_reason", reason)
                    put("banned_at", java.time.Instant.now().toString())
                }
            ) {
                filter { eq("id", userId) }
            }
        logAction("ban_user", "user", userId, "Banned: $reason")
    }

    override suspend fun unbanUser(userId: String): Result<Unit> = runCatching {
        supabaseClient.from("profiles")
            .update(
                buildJsonObject {
                    put("is_banned", false)
                    put("ban_reason", null as String?)
                    put("banned_at", null as String?)
                }
            ) {
                filter { eq("id", userId) }
            }
        logAction("unban_user", "user", userId, null)
    }

    // Roles

    override suspend fun fetchRoles(): Result<List<Role>> = runCatching {
        supabaseClient.from("roles")
            .select()
            .decodeList<Role>()
    }

    override suspend fun assignRole(userId: String, roleId: Int): Result<Unit> = runCatching {
        val adminId = supabaseClient.auth.currentUserOrNull()?.id
        supabaseClient.from("user_roles")
            .insert(
                buildJsonObject {
                    put("user_id", userId)
                    put("role_id", roleId)
                    put("assigned_by", adminId)
                }
            )
        logAction("assign_role", "role", "$userId:$roleId", null)
    }

    override suspend fun revokeRole(userId: String, roleId: Int): Result<Unit> = runCatching {
        supabaseClient.from("user_roles")
            .delete {
                filter {
                    eq("user_id", userId)
                    eq("role_id", roleId)
                }
            }
        logAction("revoke_role", "role", "$userId:$roleId", null)
    }

    // Moderation

    override suspend fun fetchModerationQueue(
        status: String?,
        limit: Int,
        offset: Int
    ): Result<List<ModerationQueueItem>> = runCatching {
        supabaseClient.from("forum_moderation_queue")
            .select {
                filter {
                    status?.let { eq("status", it) }
                }
                limit(limit.toLong())
                range(offset.toLong(), (offset + limit - 1).toLong())
                order("created_at", io.github.jan.supabase.postgrest.query.Order.DESCENDING)
            }
            .decodeList<ModerationQueueItem>()
    }

    override suspend fun resolveModerationItem(
        itemId: Int,
        resolution: ModerationResolution,
        notes: String
    ): Result<Unit> = runCatching {
        val adminId = supabaseClient.auth.currentUserOrNull()?.id
        supabaseClient.from("forum_moderation_queue")
            .update(
                buildJsonObject {
                    put("status", "resolved")
                    put("resolution", resolution.name.lowercase())
                    put("resolved_by", adminId)
                    put("resolved_at", java.time.Instant.now().toString())
                    put("resolution_notes", notes)
                }
            ) {
                filter { eq("id", itemId) }
            }
        logAction("resolve_moderation", "moderation", itemId.toString(), "${resolution.name}: $notes")
    }

    // Content

    override suspend fun deletePost(postId: Int): Result<Unit> = runCatching {
        supabaseClient.from("posts")
            .update(
                buildJsonObject { put("is_active", false) }
            ) {
                filter { eq("id", postId) }
            }
        logAction("delete_post", "post", postId.toString(), null)
    }

    override suspend fun restorePost(postId: Int): Result<Unit> = runCatching {
        supabaseClient.from("posts")
            .update(
                buildJsonObject { put("is_active", true) }
            ) {
                filter { eq("id", postId) }
            }
        logAction("restore_post", "post", postId.toString(), null)
    }

    override suspend fun deleteComment(commentId: Int): Result<Unit> = runCatching {
        supabaseClient.from("comments")
            .delete {
                filter { eq("id", commentId) }
            }
        logAction("delete_comment", "comment", commentId.toString(), null)
    }

    // Audit

    override suspend fun fetchAuditLogs(limit: Int, offset: Int): Result<List<AdminAuditLog>> = runCatching {
        supabaseClient.from("admin_audit_log")
            .select {
                limit(limit.toLong())
                range(offset.toLong(), (offset + limit - 1).toLong())
                order("created_at", io.github.jan.supabase.postgrest.query.Order.DESCENDING)
            }
            .decodeList<AdminAuditLog>()
    }

    override suspend fun logAction(
        action: String,
        resourceType: String,
        resourceId: String?,
        details: String?
    ): Result<Unit> = runCatching {
        val adminId = supabaseClient.auth.currentUserOrNull()?.id ?: return@runCatching
        supabaseClient.from("admin_audit_log")
            .insert(
                buildJsonObject {
                    put("admin_id", adminId)
                    put("action", action)
                    put("resource_type", resourceType)
                    resourceId?.let { put("resource_id", it) }
                    details?.let { put("details", it) }
                }
            )
    }

    // Auth

    override suspend fun hasAdminAccess(): Boolean {
        return try {
            val userId = supabaseClient.auth.currentUserOrNull()?.id ?: return false
            val roles = supabaseClient.from("user_roles")
                .select(Columns.raw("role_id, roles(name)")) {
                    filter { eq("user_id", userId) }
                }
                .decodeList<UserRoleWithName>()
            roles.any { it.roleName in listOf("admin", "super_admin", "moderator") }
        } catch (e: Exception) {
            false
        }
    }

    override suspend fun hasSuperAdminAccess(): Boolean {
        return try {
            val userId = supabaseClient.auth.currentUserOrNull()?.id ?: return false
            val roles = supabaseClient.from("user_roles")
                .select(Columns.raw("role_id, roles(name)")) {
                    filter { eq("user_id", userId) }
                }
                .decodeList<UserRoleWithName>()
            roles.any { it.roleName == "super_admin" }
        } catch (e: Exception) {
            false
        }
    }
}

@kotlinx.serialization.Serializable
private data class UserRoleWithName(
    @SerialName("role_id") val roleId: Int,
    @SerialName("roles") val roles: RoleName? = null
) {
    val roleName: String get() = roles?.name ?: ""
}

@kotlinx.serialization.Serializable
private data class RoleName(val name: String)

File: Sources/FoodShare/Features/Admin/Data/Repositories/SupabaseAdminRepository.swift
import Foundation
import Supabase
import FoodShareRepository

// MARK: - Supabase Admin Repository

/// Supabase implementation of AdminRepository
@MainActor
final class SupabaseAdminRepository: BaseSupabaseRepository, AdminRepository {
    private let adminAPI: AdminAPIService
    
    init(supabase: Supabase.SupabaseClient, adminAPI: AdminAPIService = .shared) {
        self.adminAPI = adminAPI
        super.init(supabase: supabase, subsystem: "com.flutterflow.foodshare", category: "AdminRepository")
    }

    // MARK: - Dashboard (Server-Side)

    func fetchDashboardStats() async throws -> AdminDashboardStats {
        let dto: DashboardStatsDTO = try await executeRPC("get_dashboard_stats", params: EmptyParams())
        
        return AdminDashboardStats(
            totalUsers: dto.totalUsers,
            activeUsers: dto.activeUsers,
            totalPosts: dto.totalPosts,
            activePosts: dto.activePosts,
            pendingReports: dto.pendingReports,
            totalMessages: dto.totalMessages,
            newUsersToday: dto.newUsersToday,
            newPostsToday: dto.newPostsToday
        )
    }

    // MARK: - Moderation Stats (Server-Side)

    func fetchModerationStats() async throws -> ModerationStatsResult {
        let dto: ModerationStatsDTO = try await executeRPC("get_moderation_stats", params: EmptyParams())
        
        return ModerationStatsResult(
            pendingCount: dto.pendingCount,
            highPriorityCount: dto.highPriorityCount,
            resolvedToday: dto.resolvedToday,
            totalReports: dto.totalReports
        )
    }

    // MARK: - Users

    func fetchUsers(
        searchQuery: String?,
        roleFilter: String?,
        limit: Int,
        offset: Int
    ) async throws -> [AdminUserProfile] {
        do {
            return try await adminAPI.getUsers(
                search: searchQuery,
                role: roleFilter,
                limit: limit,
                offset: offset
            )
        } catch {
            // Fallback to direct RPC if Edge Function is unavailable
            return try await executeRPC(
                "get_admin_users",
                params: AdminUsersParams(
                    pSearchQuery: searchQuery,
                    pRoleFilter: roleFilter,
                    pLimit: limit,
                    pOffset: offset
                )
            )
        }
    }

    func fetchUser(id: UUID) async throws -> AdminUserProfile {
        try await fetchOne(
            from: "profiles",
            select: """
                id, nickname, email, avatar_url, is_verified, is_active, created_time, last_seen_at,
                user_roles(role_id, role:roles(id, name))
            """,
            id: id.hashValue
        )
    }

    func updateUserStatus(userId: UUID, isActive: Bool) async throws {
        try await update(
            table: "profiles",
            id: userId.hashValue,
            value: ["is_active": isActive]
        )
    }

    func banUser(userId: UUID, reason: String) async throws {
        do {
            try await adminAPI.banUser(userId: userId, reason: reason)
        } catch {
            // Fallback to direct Supabase if Edge Function is unavailable
            try await supabase
                .from("profiles")
                .update(["is_active": false])
                .eq("id", value: userId)
                .execute()

            try await supabase
                .from("forum_user_warnings")
                .insert([
                    "profile_id": userId.uuidString,
                    "warning_type": "perm_ban",
                    "reason": reason,
                ])
                .execute()
        }
    }

    func unbanUser(userId: UUID) async throws {
        do {
            try await adminAPI.unbanUser(userId: userId)
        } catch {
            // Fallback to direct Supabase if Edge Function is unavailable
            try await supabase
                .from("profiles")
                .update(["is_active": true])
                .eq("id", value: userId)
                .execute()
        }
    }

    // MARK: - Roles

    func fetchRoles() async throws -> [Role] {
        let response: [Role] = try await supabase
            .from("roles")
            .select()
            .order("id", ascending: true)
            .execute()
            .value

        return response
    }

    func fetchUserRoles(userId: UUID) async throws -> [UserRole] {
        let response: [UserRole] = try await supabase
            .from("user_roles")
            .select("*, role:roles(*)")
            .eq("profile_id", value: userId)
            .execute()
            .value

        return response
    }

    func assignRole(userId: UUID, roleId: Int, grantedBy: UUID) async throws {
        try await supabase
            .from("user_roles")
            .insert([
                "profile_id": userId.uuidString,
                "role_id": String(roleId),
                "granted_by": grantedBy.uuidString
            ])
            .execute()
    }

    func revokeRole(userId: UUID, roleId: Int) async throws {
        try await supabase
            .from("user_roles")
            .delete()
            .eq("profile_id", value: userId)
            .eq("role_id", value: roleId)
            .execute()
    }

    // MARK: - Moderation

    func fetchModerationQueue(
        status: String?,
        contentType: String?,
        limit: Int,
        offset: Int,
    ) async throws -> [ModerationQueueItem] {
        var query = supabase
            .from("forum_moderation_queue")
            .select("""
                *,
                reporter:profiles!forum_moderation_queue_reporter_id_fkey(id, nickname, avatar_url),
                target_user:profiles!forum_moderation_queue_profile_id_fkey(id, nickname, avatar_url)
            """)

        if let status {
            query = query.eq("status", value: status)
        }

        if let contentType {
            query = query.eq("content_type", value: contentType)
        }

        let response: [ModerationQueueItem] = try await query
            .order("priority", ascending: false)
            .order("created_at", ascending: true)
            .range(from: offset, to: offset + limit - 1)
            .execute()
            .value

        return response
    }

    func resolveModerationItem(
        itemId: UUID,
        resolution: String,
        notes: String?,
        resolvedBy: UUID,
    ) async throws {
        struct ResolveModerationDTO: Encodable {
            let status: String
            let resolution: String
            let resolved_by: String
            let resolved_at: String
            let resolution_notes: String?
        }

        let updateData = ResolveModerationDTO(
            status: "resolved",
            resolution: resolution,
            resolved_by: resolvedBy.uuidString,
            resolved_at: Date().ISO8601Format(),
            resolution_notes: notes,
        )

        try await supabase
            .from("forum_moderation_queue")
            .update(updateData)
            .eq("id", value: itemId)
            .execute()
    }

    // MARK: - Content Management

    func deletePost(postId: Int, reason: String) async throws {
        do {
            try await adminAPI.deactivateListing(id: postId)
        } catch {
            // Fallback to direct Supabase if Edge Function is unavailable
            try await supabase
                .from("posts")
                .update(["is_active": false])
                .eq("id", value: postId)
                .execute()
        }
    }

    func restorePost(postId: Int) async throws {
        do {
            try await adminAPI.activateListing(id: postId)
        } catch {
            // Fallback to direct Supabase if Edge Function is unavailable
            try await supabase
                .from("posts")
                .update(["is_active": true])
                .eq("id", value: postId)
                .execute()
        }
    }

    func deleteComment(commentId: Int, reason: String) async throws {
        try await supabase
            .from("comments")
            .delete()
            .eq("id", value: commentId)
            .execute()
    }

    // MARK: - Audit Log

    func fetchAuditLogs(
        adminId: UUID?,
        action: String?,
        resourceType: String?,
        limit: Int,
        offset: Int,
    ) async throws -> [AdminAuditLog] {
        var query = supabase
            .from("admin_audit_log")
            .select("*, admin:profiles!admin_audit_log_admin_id_fkey(id, nickname, avatar_url)")

        if let adminId {
            query = query.eq("admin_id", value: adminId)
        }

        if let action {
            query = query.eq("action", value: action)
        }

        if let resourceType {
            query = query.eq("resource_type", value: resourceType)
        }

        let response: [AdminAuditLog] = try await query
            .order("created_at", ascending: false)
            .range(from: offset, to: offset + limit - 1)
            .execute()
            .value

        return response
    }

    func logAction(
        adminId: UUID,
        action: AdminAction,
        resourceType: AdminResourceType,
        resourceId: String?,
        metadata: [String: String]?,
        success: Bool,
        errorMessage: String?,
    ) async throws {
        struct AuditLogDTO: Encodable {
            let admin_id: String
            let action: String
            let resource_type: String
            let success: Bool
            let resource_id: String?
            let metadata: [String: String]?
            let error_message: String?
        }

        let logData = AuditLogDTO(
            admin_id: adminId.uuidString,
            action: action.rawValue,
            resource_type: resourceType.rawValue,
            success: success,
            resource_id: resourceId,
            metadata: metadata,
            error_message: errorMessage,
        )

        try await supabase
            .from("admin_audit_log")
            .insert(logData)
            .execute()
    }

    // MARK: - Authorization

    func hasAdminAccess(userId: UUID) async throws -> Bool {
        let response: [UserRole] = try await supabase
            .from("user_roles")
            .select("*, role:roles(*)")
            .eq("profile_id", value: userId)
            .execute()
            .value

        return response.contains { $0.role?.isAdmin == true }
    }

    func hasSuperAdminAccess(userId: UUID) async throws -> Bool {
        let response: [UserRole] = try await supabase
            .from("user_roles")
            .select("*, role:roles(*)")
            .eq("profile_id", value: userId)
            .execute()
            .value

        return response.contains { $0.role?.isSuperAdmin == true }
    }
}

// MARK: - DTOs

/// DTO for decoding the get_moderation_stats RPC response
private struct ModerationStatsDTO: Decodable {
    let pendingCount: Int
    let highPriorityCount: Int
    let resolvedToday: Int
    let totalReports: Int

    enum CodingKeys: String, CodingKey {
        case pendingCount = "pending_count"
        case highPriorityCount = "high_priority_count"
        case resolvedToday = "resolved_today"
        case totalReports = "total_reports"
    }
}

/// Parameters for the get_admin_users RPC
private struct AdminUsersParams: Encodable {
    let pSearchQuery: String?
    let pRoleFilter: String?
    let pLimit: Int
    let pOffset: Int

    enum CodingKeys: String, CodingKey {
        case pSearchQuery = "p_search_query"
        case pRoleFilter = "p_role_filter"
        case pLimit = "p_limit"
        case pOffset = "p_offset"
    }
}

/// DTO for decoding the get_dashboard_stats RPC response
private struct DashboardStatsDTO: Decodable {
    let totalUsers: Int
    let activeUsers: Int
    let totalPosts: Int
    let activePosts: Int
    let pendingReports: Int
    let totalMessages: Int
    let newUsersToday: Int
    let newPostsToday: Int

    enum CodingKeys: String, CodingKey {
        case totalUsers = "total_users"
        case activeUsers = "active_users"
        case totalPosts = "total_posts"
        case activePosts = "active_posts"
        case pendingReports = "pending_reports"
        case totalMessages = "total_messages"
        case newUsersToday = "new_users_today"
        case newPostsToday = "new_posts_today"
    }
}
